include/dd/Package.hpp

```diff
@@ -1274,22 +1274,21 @@ namespace dd {
             return fid.r * fid.r + fid.i * fid.i;
         }
 
-        dd::fp fidelityOfMeasurementOutcomes(const vEdge& e, const std::vector<dd::fp>& probs) {
+        dd::fp fidelityOfMeasurementOutcomes(const vEdge& e, const ProbabilityVector& probs) {
             if (e.w.approximatelyZero()) {
                 return 0.;
             }
-            const auto nq = e.p->v + 1;
-            if (probs.size() != (1U << nq)) {
-                throw std::runtime_error("Mismatch in sizes of DD and probability vector in fidelity function.");
-            }
-
             return fidelityOfMeasurementOutcomesRecursive(e, probs, 0);
         }
 
-        dd::fp fidelityOfMeasurementOutcomesRecursive(const vEdge& e, const std::vector<dd::fp>& probs, const std::size_t i) {
+        dd::fp fidelityOfMeasurementOutcomesRecursive(const vEdge& e, const ProbabilityVector& probs, const std::size_t i) {
             const auto topw = dd::ComplexNumbers::mag(e.w);
             if (e.isTerminal()) {
-                return topw * std::sqrt(probs[i]);
+                if (auto it = probs.find(i); it != probs.end()) {
```

Comment:
The keys `prop` have type `std::uint_least64_t` whereas the `find` method is called with a `std::size_t`. Make `ProbabilityVector` a `std::unordered_map<std::size_t, fp>`?